program/src/processor.rs

```diff
@@ -5313,6 +5318,286 @@ impl Processor {
+            specified Mango vaults that were passed in first. We also made sure all fields except
+            `amount` are correct. At the end of the instruction we take the diff in `amount` and
+            assign changes to `MangoAccount`. We ensure all other fields stayed the same during CPI.
+         */
+        for ai in cpi_ais {
+            // ignore accounts that are not writable
+            if !ai.is_writable {
+                // TODO - verify that this field cannot change over course of a tx
+                continue;
+            }
+
+            // Serum dex accounts cannot be sent writable for safety reasons
+            check!(ai.owner != &mango_group.dex_program_id, MangoErrorCode::InvalidAccount)?;
+
+            // If an SPL token Account is owned by mango signer key, make sure it's one of the bank_sets
+            if ai.owner == &spl_token::id() && ai.data_len() == Account::LEN {
+                // let token_account = Account::unpack(&ai.try_borrow_data()?)?;
+                let token_account = TokenAccount::load_checked(ai)?;
+                if &token_account.0.owner == &mango_group.signer_key {
+                    bank_sets
+                        .iter()
+                        .find(|bank_set| bank_set.vault_ai.key == ai.key)
+                        .ok_or(throw_err!(MangoErrorCode::InvalidVault))?;
+                }
+            }
+        }
+
+        // Unpack the open orders accounts and compute MangoAccount health before CPI
+        let open_orders_ais =
+            mango_account.checked_unpack_open_orders(&mango_group, packed_open_orders_ais)?;
+        let mut health_cache = HealthCache::new(active_assets);
+        health_cache.init_vals_with_orders_vec(
+            &mango_group,
+            &mango_cache,
+            &mango_account,
+            &open_orders_ais,
+        )?;
+        let pre_health = health_cache.get_health(&mango_group, HealthType::Init);
+
+        // update the being_liquidated flag
+        if mango_account.being_liquidated {
+            if pre_health >= ZERO_I80F48 {
+                mango_account.being_liquidated = false;
+            } else {
+                return Err(throw_err!(MangoErrorCode::BeingLiquidated));
+            }
+        }
+
+        // This means health must only go up
+        let reduce_only = pre_health < ZERO_I80F48;
+
+        // Prepare and invoke CPI
+        let cpi_account_metas: Vec<AccountMeta> = cpi_ais
+            .iter()
+            .map(|cpi_ai| {
+                if cpi_ai.is_writable {
+                    AccountMeta::new(*cpi_ai.key, cpi_ai.is_signer)
+                } else {
+                    AccountMeta::new_readonly(*cpi_ai.key, cpi_ai.is_signer)
+                }
+            })
+            .collect();
+        let mut cpi_account_infos: Vec<AccountInfo> = cpi_ais.iter().map(|ai| ai.clone()).collect();
+        cpi_account_infos.push(cpi_prog_ai.clone());
```

Comment:
why's the program_ai added?

Reply:
requirement of Solana runtime that you pass in that account during CPI. If you notice, we do the same for every other CPI